biogeophys/FatesPlantRespPhotosynthMod.F90

```diff
@@ -399,504 +413,608 @@ subroutine FatesPlantRespPhotosynthDrive (nsites, sites,bc_in,bc_out,dtime)
+                                  
+                               else
+                                  par_per_sunla = 0._r8
+                               end if
+                               
+                               if(((laisha*currentPatch%canopy_area_profile(cl,ft,iv)) >nearzero) .and. &
+                                    (currentPatch%ed_parsha_z(cl,ft,iv)>nearzero)) then
+                                  par_per_shala = currentPatch%ed_parsha_z(cl,ft,iv) / &
+                                       (laisha*currentPatch%canopy_area_profile(cl,ft,iv))
+                               else
+                                  par_per_shala = 0._r8
+                               end if
+                               
+                               fsun = currentPatch%f_sun(cl,ft,iv)
+                               elai_layer = currentPatch%elai_profile(cl,ft,iv)
+                               
+                            else
+                               
+                               if(cohort_layer_elai(iv) > nearzero .and. currentPatch%solar_zenith_flag) then
+                                  
+                                  call FatesGetCohortAbsRad(currentPatch, currentCohort, ipar, &
+                                       cohort_vaitop(iv), cohort_vaibot(iv), cohort_elai, cohort_esai, &
+                                       rb_abs, rd_abs, rb_abs_leaf, rd_abs_leaf, fsun)
+                                  
+                                  ! rd_abs_leaf: Watts of diffuse light absorbed by leaves over this
+                                  !              depth interval and ground footprint (m2)
+                                  ! rd_abs_leaf*fsun  Watts of diffuse light absorbed by sunlit leaves
+                                  !                   over this depth interval and ground footprint (m2)
+                                  ! rb_abs_leaf       Watts of beam absorbed by sunlit leaves over this
+                                  !                   depth interval and ground footprint (m2)
+                                  ! cohort_layer_elai*fsun       Leaf area in sunlight within this interval and ground footprint
+                                  ! cohort_layer_elai*(1-fsun)   Leaf area in shade within this interval and ground footprint
+
+                                  if(fsun>nearzero) then
+                                     par_per_sunla = (rd_abs_leaf*fsun + rb_abs_leaf) / (fsun*cohort_layer_elai(iv))
+                                  else
+                                     par_per_sunla = 0._r8
+                                  end if
+                                  par_per_shala = rd_abs_leaf*(1._r8-fsun) / ((1._r8 - fsun)*cohort_layer_elai(iv))
+                                  
+                               else
+                                  
+                                  par_per_sunla = 0._r8
+                                  par_per_shala = 0._r8
```

Comment:
I could run some tests to see if the atmospheric boundary conditions for diffuse are ever non-zero for cos(zen) = 0.  If this does happen, we could consider doing something here, or in a new PR.  I suppose we use the zenith angle as the filter for performing the solve, because this happens before the downwelling diffuse boundary conditions is known, via a normalized calculation. 
Moving forward, for twilight, we could decide to solve if either of two conditions is true, if the cos(zen)>0, or if the previous atmospheric diffuse>0.  Aside from wasting a little time, there is no liability to running the solver and not using it if there is no diffuse light.  

If the diffuse scattering coefficients didn't change over the course of the day due to ice/wetness conditions, we could split the solver and just run the normalized diffuse calculations once per day when the canopy changes, and use that for all diffuse boundary conditions, but changing wetness/ice is a thing.

Reply:
I think this depends on the atmospheric boundary conditions, gridded ones may or may not have, whereas tower-based ones may some irradiance during twilight, at least if they are from high latitudes. In any case, I think this makes more sense for a future PR, because this is not exclusive of two-stream.

For reference, in ED2 we use two criteria to test whether or not to call the radiation driver: if zen < 89.9° or so, then it is daytime. If 89.9° ≤ zen < 96° (end of civil twilight) *and* total incoming irradiance is greater than 0.5 W m<sup>−2</sup>, then it is twilight (diffuse only), otherwise it assumes night and skips the radiation solver altogether.